File: TPs/TP1/3_Is/Is.java
//Samuel Fragoso - 902665
import java.util.Scanner;

class Is{
    /**
    * Verifica se a string e composta apenas por vogais.
    * @param s String a ser avaliada.
    * @return true se tiver apenas vogais, false caso contrario.
    */
    static boolean isVogal(String s){
        boolean resp = true;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
               if (c != 'a' && c != 'e' && c != 'i' && c != 'o' && c != 'u' &&
                   c != 'A' && c != 'E' && c != 'I' && c != 'O' && c != 'U'){
                   resp = false;
               } 
        }
        return resp;
    }

    /**
     * Verifica se a string e composta apenas por vogais.
     * @param s String a ser avaliada.
     * @return true se tiver apenas vogais, false caso contrario.
     */
    static boolean isConsoante(String s){
        boolean resp = true;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            boolean isLetra = (c >='a' && c <= 'z') || (c >= 'A' && c <= 'Z');

            boolean isVogal = (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
                               c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U');
            
            if (!isLetra || isVogal) {
            return false; //se não for letra ou se for vogal, já não é só consoante
        }                   

        }
        return resp;
    }
    /**
     * Verifica se a string e um numero inteiro.
     * @param s String a ser avaliada.
     * @return true se for inteiro, false caso contrario.
     */
    static boolean isInteiro(String s){
        boolean resp = true;
        for(int i=0; i<s.length();i++){
            char c = s.charAt(i);
            if(c < '0' || c > '9'){
                resp = false;
            }
        }
        return resp;
    }

    /**
     * Verifica se a string e um numero real.
     * @param s String a ser avaliada.
     * @return true se for real, false caso contrario.
     */
    static boolean isReal(String s){
        int virgulasOuPontos = 0;
        boolean resp = true;
        for(int i=0; i<s.length();i++){
            char c = s.charAt(i);
            if (c >= '0' && c <= '9') {
                continue; 
            } else if (c == ',' || c == '.') {
                virgulasOuPontos++;
            } else {
                return false; //caractere invalido (letra, etc)
            }
        }
        return virgulasOuPontos <= 1;
    }

    //metodo auxiliar para converter boolean em SIM/NAO
    static String simNao(boolean b){
        return b ? "SIM" : "NAO";
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in, "UTF-8");
        while (sc.hasNextLine()){
        String linha = sc.nextLine();
        if (linha.length() == 3 && linha.charAt(0) == 'F' && 
            linha.charAt(1) == 'I' && linha.charAt(2) == 'M') break;

        // Imprime os 4 resultados na mesma linha 
        System.out.print(simNao(isVogal(linha)) + " ");
        System.out.print(simNao(isConsoante(linha)) + " ");
        System.out.print(simNao(isInteiro(linha)) + " ");
        System.out.println(simNao(isReal(linha)));    
        }
        sc.close();
    }
}

File: TPs/TP1/10_Is_recursivo/Is.c
//Samuel Fragoso
#include <stdio.h>
#include <stdlib.h>

/**
 * Funcao auxiliar para converter para minuscula manualmente.
 */
char toLower(char c){
    if(c >='A' && c <= 'Z') return c + 32;
    return c;
}

/**
 * Verifica recursivamente se a string contem apenas vogais.
 */
int apenasVogais(char* s, int i){
    if (s[i] == '\0' || s[i] == '\n' || s[i] == '\r') return 1;

    char c = toLower(s[i]);
    if(c == 'a' || c=='e' || c=='i' || c=='o'||c=='u'){
        return apenasVogais(s, i+1);
    }
    return 0;
}

/**
 * Verifica recursivamente se a string contem apenas consoantes.
 */
int apenasConsoantes(char* s, int i){
    if(s[i] == '\0' || s[i] == '\n' || s[i] == '\r') return 1;

    char c = toLower(s[i]);
    if ((c >= 'a' && c <= 'z') && !(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u')) {
        return apenasConsoantes(s, i+1);
    }   
    return 0;
}

/**
 * Verifica recursivamente se a string eh um numero inteiro.
 */
int isInteiro(char* s, int i){
    if(s[i] == '\0' || s[i] == '\n' || s[i] == '\r') return 1;

    if(s[i]>='0' && s[i]<='9'){
        return isInteiro(s, i+1);
    }
    return 0;
}

/**
 * Verifica recursivamente se a string eh um numero real.
 * @param pontos Contador de pontos ou virgulas encontrados.
 */
int isReal(char* s, int i, int pontos) {
    if (s[i] == '\0' || s[i] == '\n' || s[i] == '\r') return 1;
    
    if (s[i] >= '0' && s[i] <= '9') {
        return isReal(s, i + 1, pontos);
    } else if ((s[i] == '.' || s[i] == ',') && pontos == 0) {
        return isReal(s, i + 1, pontos + 1);
    }
    return 0;
}

/**
 * Verifica se a string eh "FIM".
 */
int isFim(char* s) {
    return (s[0] == 'F' && s[1] == 'I' && s[2] == 'M' && (s[3] == '\0' || s[3] == '\n' || s[3] == '\r'));
}

/**
* Metodo principal para leitura e exibicao dos resultados.
*/
int main(){
    char linha[5000];
    while(fgets(linha, 5000, stdin)!=NULL){
        if(isFim(linha)) break;

        printf("%s ", apenasVogais(linha, 0) ? "SIM" : "NAO");
        printf("%s ", apenasConsoantes(linha, 0) ? "SIM" : "NAO");
        printf("%s ", isInteiro(linha, 0) ? "SIM" : "NAO");
        printf("%s\n", isReal(linha, 0, 0) ? "SIM" : "NAO");
    }
    return 0;
}